ros/src/computing/perception/detection/lidar_tracker/packages/lidar_imm_ukf_pda_track/nodes/imm_ukf_pda/ukf.cpp

```diff
@@ -324,20 +408,288 @@ void UKF::predictionIMMUKF(const double dt)
+    p_ctrv_ = beta_ctrv_zero * p_pre_ctrv +
+              (1 - beta_ctrv_zero) * (p_pre_ctrv - k_ctrv_ * s_ctrv_ * k_ctrv_.transpose()) +
+              k_ctrv_ * sigma_p_ctrv * k_ctrv_.transpose();
+    p_rm_ = beta_rm_zero * p_pre_rm + (1 - beta_rm_zero) * (p_pre_rm - k_rm_ * s_rm_ * k_rm_.transpose()) +
+            k_rm_ * sigma_p_rm * k_rm_.transpose();
+  }
+  else
+  {
+    p_cv_ = p_pre_cv - k_cv_ * s_cv_ * k_cv_.transpose();
+    p_ctrv_ = p_pre_ctrv - k_ctrv_ * s_ctrv_ * k_ctrv_.transpose();
+    p_rm_ = p_pre_rm - k_rm_ * s_rm_ * k_rm_.transpose();
+  }
+
+  Eigen::VectorXd max_det_z;
+  Eigen::MatrixXd max_det_s;
+
+  findMaxZandS(max_det_z, max_det_s);
+  double Vk = M_PI * sqrt(gating_thres * max_det_s.determinant());
+
+  double lambda_cv, lambda_ctrv, lambda_rm;
+  if (num_meas != 0)
+  {
+    lambda_cv =
+        (1 - gate_probability * detection_probability) / pow(Vk, num_meas) +
+        detection_probability * pow(Vk, 1 - num_meas) * e_cv_sum / (num_meas * sqrt(2 * M_PI * s_cv_.determinant()));
+    lambda_ctrv = (1 - gate_probability * detection_probability) / pow(Vk, num_meas) +
+                  detection_probability * pow(Vk, 1 - num_meas) * e_ctrv_sum /
+                      (num_meas * sqrt(2 * M_PI * s_ctrv_.determinant()));
+    lambda_rm =
+        (1 - gate_probability * detection_probability) / pow(Vk, num_meas) +
+        detection_probability * pow(Vk, 1 - num_meas) * e_rm_sum / (num_meas * sqrt(2 * M_PI * s_rm_.determinant()));
+  }
+  else
+  {
+    lambda_cv = (1 - gate_probability * detection_probability);
+    lambda_ctrv = (1 - gate_probability * detection_probability);
+    lambda_rm = (1 - gate_probability * detection_probability);
+  }
+
+  lambda_vec.push_back(lambda_cv);
+  lambda_vec.push_back(lambda_ctrv);
+  lambda_vec.push_back(lambda_rm);
+}
+
+void UKF::updateLikelyMeasurementForCTRV(const std::vector<autoware_msgs::DetectedObject>& object_vec)
+{
+  double num_meas = object_vec.size();
+  std::vector<double> e_ctrv_vec;
+  std::vector<Eigen::VectorXd> meas_vec;
+
+  for (size_t i = 0; i < num_meas; i++)
+  {
+    Eigen::VectorXd meas = Eigen::VectorXd(2);
+    meas(0) = object_vec[i].pose.position.x;
+    meas(1) = object_vec[i].pose.position.y;
+    meas_vec.push_back(meas);
+    Eigen::VectorXd diff_ctrv = meas - z_pred_ctrv_;
+    double e_ctrv = exp(-0.5 * diff_ctrv.transpose() * s_ctrv_.inverse() * diff_ctrv);
+    e_ctrv_vec.push_back(e_ctrv);
+  }
+  // for noise estimation
+  if (num_meas != 0)
+  {
+    std::vector<double>::iterator max_ctrv_iter = std::max_element(e_ctrv_vec.begin(), e_ctrv_vec.end());
+    int max_ctrv_ind = std::distance(e_ctrv_vec.begin(), max_ctrv_iter);
+    ctrv_meas_ = meas_vec[max_ctrv_ind];
+  }
+}
+
+void UKF::updateSUKF(const std::vector<autoware_msgs::DetectedObject>& object_vec)
+{
+  if (object_vec.size() == 0)
+  {
+    std::cout << "no update for this frame since there are no associated mesurements" << std::endl;
```

Comment:
Is it better to use ROS API instead of standard output?
http://wiki.ros.org/roscpp/Overview/Logging